File: examples/jvm/src/main/java/com/example/chatbot/ChatCore.java
package com.example.chatbot;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.x.chatxdk.Chat;
import com.x.chatxdk.StreamDecryptor;
import com.x.chatxdk.StreamEncryptor;
import com.x.chatxdk.Types.*;

import java.io.ByteArrayOutputStream;
import java.util.Arrays;
import java.util.Base64;
import java.util.List;
import java.util.Map;

/**
 * Crypto core for the JVM chat-xdk example bot.
 *
 * <p>A thin, network-free wrapper around the {@link Chat} binding.
 * Everything that touches the SDK lives here so it can be unit-tested directly
 * (see {@code ChatCoreTest}). The four core feature touchpoints are all here:
 *
 * <ul>
 *   <li>key management     -&gt; loadKeys / setIdentity / generateAndRegister</li>
 *   <li>conversation keys  -&gt; prepareConversationKeyChange / decryptConversationKey</li>
 *   <li>message encryption -&gt; encryptReply</li>
 *   <li>event decryption   -&gt; decryptBatch (decryptEvents) and decryptOne (decryptEvent)</li>
 * </ul>
 */
public final class ChatCore implements AutoCloseable {

    private static final ObjectMapper MAPPER = new ObjectMapper();

    private final Chat chat = new Chat();
    private String signingKeyVersion = "1";

    public String signingKeyVersion() {
        return signingKeyVersion;
    }

    // -- Key management -----------------------------------------------------

    /** Import an existing base64 private-key blob (identity[+signing])
     * together with the key version it was registered under. */
    public void loadKeys(String privateKeysB64, String signingKeyVersion) {
        chat.importKeys(Base64.getDecoder().decode(privateKeysB64), signingKeyVersion);
        this.signingKeyVersion = signingKeyVersion;
    }

    /**
     * Set the session identity once; every encrypt/prepare call below then
     * signs as this user without passing a sender id per call.
     */
    public void setIdentity(String userId) {
        chat.setIdentity(userId, signingKeyVersion);
    }

    /** Generated identity: the registration payload + the private blob to persist. */
    public record Generated(PublicKeyRegistrationPayload registration, String privateKeysB64) {}

    /** Generate a fresh identity to register with the X API. */
    public Generated generateAndRegister() throws Exception {
        PublicKeyRegistrationPayload payload = chat.generateKeypairs();
        byte[] exported = chat.exportKeys();
        String blob = exported == null ? "" : Base64.getEncoder().encodeToString(exported);
        return new Generated(payload, blob);
    }

    public PublicKeys publicKeys() throws Exception {
        return chat.getPublicKeys();
    }

    // -- Conversation keys --------------------------------------------------

    public PreparedConversationChange prepareConversationKeyChange(
            List<PublicKeyInput> publicKeys, String conversationId) throws Exception {
        ConversationKeyChangeParams params = new ConversationKeyChangeParams(publicKeys);
        params.conversationId = conversationId;
        return chat.prepareConversationKeyChange(params);
    }

    /** ECIES-decrypt one conversation key -&gt; raw 32-byte key. */
    public byte[] decryptConversationKey(String encryptedKeyB64) {
        return chat.decryptConversationKey(encryptedKeyB64);
    }

    // -- Decryption: the two paths -----------------------------------------

    /** Batch path — used on initial conversation load. */
    public DecryptEventsResult decryptBatch(List<String> eventsB64, List<SigningKeyEntry> signingKeys) throws Exception {
        return chat.decryptEvents(eventsB64, signingKeys);
    }

    /** Single-event path — used for each new event after the initial load. */
    public JsonNode decryptOne(String eventB64, Map<String, byte[]> conversationKeys, List<SigningKeyEntry> signingKeys) throws Exception {
        return chat.decryptEvent(eventB64, conversationKeys, signingKeys);
    }

    // -- Message encryption -------------------------------------------------

    /** Fields the X API expects for an encrypted message send. */
    public record SendBody(String messageId, String encodedMessageCreateEvent, String encodedMessageEventSignature) {}

    /**
     * Encrypt + sign a fresh message ({@code encryptMessage}); returns fields
     * ready for the X API send. The sender comes from {@link #setIdentity}.
     */
    public SendBody encryptReply(String conversationId, String text, byte[] conversationKey, String conversationKeyVersion) throws Exception {
        return encryptReply(conversationId, text, conversationKey, conversationKeyVersion, null, null, null, null, null);
    }

    /**
     * Encrypt + sign a message, returning fields ready for the X API send.
     *
     * <p>Without {@code replyToEvent} this sends a fresh message via
     * {@code encryptMessage}; with it, the SDK's {@code encryptReply} builds a
     * <em>threaded</em> reply whose preview is derived from that raw signed
     * event ({@code replyToCkces} carries the key-change events for the
     * original's key version when it differs from this reply's version).
     * {@code entities} are byte-range descriptors (start, end, type);
     * {@code attachments} are attachment descriptors (e.g. a media reference);
     * {@code ttlMsec} makes the message disappear after the given lifetime.
     */
    public SendBody encryptReply(
            String conversationId,
            String text,
            byte[] conversationKey,
            String conversationKeyVersion,
            String replyToEvent,
            List<EntityDescriptor> entities,
            List<AttachmentDescriptor> attachments,
            Long ttlMsec,
            List<String> replyToCkces)
            throws Exception {
        SendPayload payload;
        if (replyToEvent == null) {
            EncryptMessageParams params = new EncryptMessageParams(conversationId, text);
            params.conversationKey = conversationKey;
            params.conversationKeyVersion = conversationKeyVersion;
            params.entities = entities;
            params.attachments = attachments;
            params.ttlMsec = ttlMsec;
            payload = chat.encryptMessage(params);
        } else {
            EncryptReplyParams params = new EncryptReplyParams(conversationId, text, replyToEvent);
            params.conversationKey = conversationKey;
            params.conversationKeyVersion = conversationKeyVersion;
            params.replyToCkces = replyToCkces;
            params.entities = entities;
            params.attachments = attachments;
            params.ttlMsec = ttlMsec;
            payload = chat.encryptReply(params);
        }
        // The SDK generates the message id and returns it in the payload.
        return new SendBody(payload.messageId, payload.encryptedContent, payload.encodedEventSignature);
    }

    /** Encrypt + sign a reaction add/remove targeting a raw event
     * (the conversation id and target sequence id are derived from it). */
    public SendBody encryptReaction(
            boolean add,
            String targetEventB64,
            String emoji,
            byte[] conversationKey,
            String conversationKeyVersion)
            throws Exception {
        EncryptReactionParams params = new EncryptReactionParams(targetEventB64, emoji);
        params.conversationKey = conversationKey;
        params.conversationKeyVersion = conversationKeyVersion;
        SendPayload payload = add ? chat.encryptAddReaction(params) : chat.encryptRemoveReaction(params);
        // The SDK generates the message id and returns it in the payload.
        return new SendBody(payload.messageId, payload.encryptedContent, payload.encodedEventSignature);
    }

    // -- Group management -----------------------------------------------------

    /** Prepare a group creation: fresh key + the two required signatures. */
    public PreparedConversationChange prepareGroupCreate(
            List<PublicKeyInput> publicKeys,
            String conversationId,
            List<String> memberIds,
            List<String> adminIds)
            throws Exception {
        return chat.prepareGroupCreate(
                new GroupCreateParams(publicKeys, conversationId, memberIds, adminIds));
    }

    /** Prepare a member add: rotated key + the two required signatures. */
    public PreparedConversationChange prepareGroupMembersChange(
            List<PublicKeyInput> publicKeys,
            String conversationId,
            List<String> newMemberIds,
            List<String> currentMemberIds,
            List<String> currentAdminIds)
            throws Exception {
        return chat.prepareGroupMembersChange(new GroupMembersChangeParams(
                publicKeys, conversationId, newMemberIds, currentMemberIds, currentAdminIds, List.of()));
    }

    // -- Media streaming -----------------------------------------------------

    private static final int MEDIA_CHUNK = 1024 * 1024;

    /**
     * Encrypt a media blob with the incremental stream API.
     *
     * <p>Feeding fixed-size chunks through {@code push} keeps memory bounded no
     * matter how large the file is; {@code finish} emits the final frame that
     * seals the stream (decryption fails without it).
     */
    public byte[] encryptMedia(byte[] plaintext, byte[] conversationKey) throws Exception {
        try (StreamEncryptor enc = chat.streamEncryptor(conversationKey)) {
            ByteArrayOutputStream out = new ByteArrayOutputStream();
            for (int offset = 0; offset < plaintext.length; offset += MEDIA_CHUNK) {
                int end = Math.min(offset + MEDIA_CHUNK, plaintext.length);
                out.write(enc.push(Arrays.copyOfRange(plaintext, offset, end)));
            }
            out.write(enc.finish());
            return out.toByteArray();
        }
    }

    /**
     * Decrypt a media blob with the incremental stream API.
     *
     * <p>{@code finish} throws if the stream was truncated, so plaintext from
     * {@code push} must not be treated as complete until it succeeds.
     */
    public byte[] decryptMedia(byte[] ciphertext, byte[] conversationKey) throws Exception {
        try (StreamDecryptor dec = chat.streamDecryptor(conversationKey)) {
            ByteArrayOutputStream out = new ByteArrayOutputStream();
            for (int offset = 0; offset < ciphertext.length; offset += MEDIA_CHUNK) {
                int end = Math.min(offset + MEDIA_CHUNK, ciphertext.length);
                out.write(dec.push(Arrays.copyOfRange(ciphertext, offset, end)));
            }
            out.write(dec.finish());
            return out.toByteArray();
        }
    }

    // -- Generic helpers (handy for metadata + tests) -----------------------

    public String encrypt(String plaintext, byte[] conversationKey) {
        return chat.encrypt(plaintext, conversationKey);
    }

    public String decrypt(String ciphertextB64, byte[] conversationKey) {
        return chat.decrypt(ciphertextB64, conversationKey);
    }

    /** Pull the plain text out of a decrypted Message event, or null. */
    public static String messageText(JsonNode event) {
        if (event == null || !"Message".equals(event.path("type").asText())) {
            return null;
        }
        JsonNode text = event.path("content").path("text");
        return text.isMissingNode() || text.isNull() ? null : text.asText();
    }

    /**
     * Map a prepared conversation change into the X API request shape.
     *
     * <p>Works for 1:1 key changes (one signature) and group create / member add
     * (two signatures). {@code signingPublicKey} is the sender's own signing key,
     * which the API expects alongside each signature. A conversation-key change's
     * {@code signature_payload} is withheld (it embeds the plaintext key), so it
     * is included only when non-empty.
     */
    public static ObjectNode prepToRequest(PreparedConversationChange prep, String signingPublicKey) {
        ObjectNode body = MAPPER.createObjectNode();
        body.put("conversation_key_version", prep.conversationKeyVersion);
        ArrayNode participantKeys = body.putArray("conversation_participant_keys");
        for (EncryptedKeyForRecipient pk : prep.participantKeys) {
            ObjectNode entry = participantKeys.addObject();
            entry.put("user_id", pk.userId);
            entry.put("encrypted_conversation_key", pk.encryptedKey);
            entry.put("public_key_version", pk.publicKeyVersion);
        }
        ArrayNode signatures = body.putArray("action_signatures");
        for (ActionSignature sig : prep.actionSignatures) {
            ObjectNode entry = signatures.addObject();
            entry.put("message_id", sig.messageId);
            entry.put("encoded_message_event_detail", sig.encodedMessageEventDetail);
            ObjectNode inner = entry.putObject("message_event_signature");
            inner.put("signature", sig.signature);
            inner.put("signature_version", sig.signatureVersion);
            inner.put("public_key_version", sig.publicKeyVersion);
            inner.put("signing_public_key", signingPublicKey);
            if (sig.signaturePayload != null && !sig.signaturePayload.isEmpty()) {
                entry.put("signature_payload", sig.signaturePayload);
            }
        }
        return body;
    }

    @Override
    public void close() {
        chat.close();
    }
}

File: examples/dotnet/ChatBot/ChatCore.cs
using System.Text.Json;
using ChatXdk;

namespace ChatBot;

/// <summary>
/// Crypto core for the .NET chat-xdk example bot.
///
/// A thin, network-free wrapper around the <see cref="ChatXdk.Chat"/>
/// binding. Everything that touches the SDK lives here so it can be unit-tested
/// directly (see ChatBot.Tests). The four core feature touchpoints are all here:
///
/// <list type="bullet">
///   <item>key management     -> LoadKeys / SetIdentity / GenerateAndRegister</item>
///   <item>conversation keys  -> PrepareConversationKeyChange / DecryptConversationKey</item>
///   <item>message encryption -> EncryptReply</item>
///   <item>event decryption   -> DecryptBatch (DecryptEvents) and DecryptOne (DecryptEvent)</item>
/// </list>
/// </summary>
public sealed class ChatCore : IDisposable
{
    private readonly Chat _chat = new();
    public string SigningKeyVersion { get; private set; } = "1";

    // -- Key management -----------------------------------------------------

    /// <summary>Import an existing base64 private-key blob (identity[+signing])
    /// together with the key version it was registered under.</summary>
    public void LoadKeys(string privateKeysB64, string signingKeyVersion = "1")
    {
        _chat.ImportKeys(Convert.FromBase64String(privateKeysB64), signingKeyVersion);
        SigningKeyVersion = signingKeyVersion;
    }

    /// <summary>
    /// Set the session identity once; every encrypt/prepare call below then
    /// signs as this user without passing a sender id per call.
    /// </summary>
    public void SetIdentity(string userId) => _chat.SetIdentity(userId, SigningKeyVersion);

    /// <summary>Generate a fresh identity. Returns the registration payload to
    /// POST to the X API plus the exported private blob (base64) to persist.</summary>
    public (PublicKeyRegistrationPayload Registration, string PrivateKeysB64) GenerateAndRegister()
    {
        var payload = _chat.GenerateKeypairs();
        var exported = _chat.ExportKeys() ?? Array.Empty<byte>();
        return (payload, Convert.ToBase64String(exported));
    }

    public PublicKeys PublicKeys() => _chat.GetPublicKeys();

    // -- Conversation keys --------------------------------------------------

    public PreparedConversationChange PrepareConversationKeyChange(
        IReadOnlyList<PublicKeyInput> publicKeys, string? conversationId = null)
        => _chat.PrepareConversationKeyChange(new ConversationKeyChangeParams(publicKeys)
        {
            ConversationId = conversationId,
        });

    /// <summary>ECIES-decrypt one conversation key -> raw 32-byte key.</summary>
    public byte[] DecryptConversationKey(string encryptedKeyB64)
        => _chat.DecryptConversationKey(encryptedKeyB64);

    // -- Decryption: the two paths -----------------------------------------

    /// <summary>Batch path — used on initial conversation load.</summary>
    public DecryptEventsResult DecryptBatch(IEnumerable<string> eventsB64, IEnumerable<SigningKeyEntry>? signingKeys = null)
        => _chat.DecryptEvents(eventsB64, signingKeys);

    /// <summary>Single-event path — used for each new event after the initial load.</summary>
    public JsonElement DecryptOne(string eventB64, Dictionary<string, byte[]> conversationKeys, IEnumerable<SigningKeyEntry>? signingKeys = null)
        => _chat.DecryptEvent(eventB64, conversationKeys, signingKeys);

    // -- Message encryption -------------------------------------------------

    /// <summary>
    /// Encrypt + sign a message; returns fields ready for the X API send.
    /// The sender comes from <see cref="SetIdentity"/>. Without
    /// <paramref name="replyToEvent"/> this sends a fresh message via
    /// <c>EncryptMessage</c>; with it, the SDK's <c>EncryptReply</c> builds a
    /// *threaded* reply whose preview is derived from that raw signed event.
    /// <paramref name="entities"/> are (start, end, type) byte ranges;
    /// <paramref name="attachments"/> are attachment descriptors (e.g. a media
    /// reference); <paramref name="ttlMsec"/> makes the message disappear
    /// after the given lifetime.
    /// </summary>
    public SendBody EncryptReply(
        string conversationId,
        string text,
        byte[] conversationKey,
        string conversationKeyVersion,
        string? replyToEvent = null,
        IReadOnlyList<EntityDescriptor>? entities = null,
        IReadOnlyList<AttachmentDescriptor>? attachments = null,
        long? ttlMsec = null,
        IReadOnlyList<string>? replyToCkces = null)
    {
        var payload = replyToEvent is null
            ? _chat.EncryptMessage(new EncryptMessageParams(conversationId, text)
            {
                ConversationKey = conversationKey,
                ConversationKeyVersion = conversationKeyVersion,
                Entities = entities,
                Attachments = attachments,
                TtlMsec = ttlMsec,
            })
            : _chat.EncryptReply(new EncryptReplyParams(conversationId, text, replyToEvent)
            {
                ConversationKey = conversationKey,
                ConversationKeyVersion = conversationKeyVersion,
                // Key-change events for the original's key version, when it
                // differs from this reply's version.
                ReplyToCkces = replyToCkces,
                Entities = entities,
                Attachments = attachments,
                TtlMsec = ttlMsec,
            });
        // The SDK generates the message id and returns it in the payload.
        return new SendBody(payload.MessageId, payload.EncryptedContent, payload.EncodedEventSignature);
    }

    /// <summary>Encrypt + sign a reaction add/remove targeting a raw event
    /// (the conversation id and target sequence id are derived from it).</summary>
    public SendBody EncryptReaction(
        bool add,
        string targetEventB64,
        string emoji,
        byte[] conversationKey,
        string conversationKeyVersion)
    {
        var parameters = new EncryptReactionParams(targetEventB64, emoji)
        {
            ConversationKey = conversationKey,
            ConversationKeyVersion = conversationKeyVersion,
        };
        var payload = add ? _chat.EncryptAddReaction(parameters) : _chat.EncryptRemoveReaction(parameters);
        // The SDK generates the message id and returns it in the payload.
        return new SendBody(payload.MessageId, payload.EncryptedContent, payload.EncodedEventSignature);
    }

    // -- Group management -----------------------------------------------------

    /// <summary>Prepare a group creation: fresh key + the two required signatures.</summary>
    public PreparedConversationChange PrepareGroupCreate(
        IReadOnlyList<PublicKeyInput> publicKeys,
        string conversationId,
        IReadOnlyList<string> memberIds,
        IReadOnlyList<string> adminIds)
        => _chat.PrepareGroupCreate(new GroupCreateParams(publicKeys, conversationId, memberIds, adminIds));

    /// <summary>Prepare a member add: rotated key + the two required signatures.</summary>
    public PreparedConversationChange PrepareGroupMembersChange(
        IReadOnlyList<PublicKeyInput> publicKeys,
        string conversationId,
        IReadOnlyList<string> newMemberIds,
        IReadOnlyList<string> currentMemberIds,
        IReadOnlyList<string> currentAdminIds)
        => _chat.PrepareGroupMembersChange(new GroupMembersChangeParams(
            publicKeys,
            conversationId,
            newMemberIds,
            currentMemberIds,
            currentAdminIds,
            Array.Empty<string>()));

    // -- Media streaming -----------------------------------------------------

    private const int MediaChunk = 1024 * 1024;

    /// <summary>
    /// Encrypt a media blob with the incremental stream API.
    ///
    /// Feeding fixed-size chunks through <c>Push</c> keeps memory bounded no
    /// matter how large the file is; <c>Finish</c> emits the final frame that
    /// seals the stream (decryption fails without it).
    /// </summary>
    public byte[] EncryptMedia(byte[] plaintext, byte[] conversationKey)
    {
        using var enc = _chat.StreamEncryptor(conversationKey);
        using var output = new MemoryStream();
        for (var offset = 0; offset < plaintext.Length; offset += MediaChunk)
        {
            var len = Math.Min(MediaChunk, plaintext.Length - offset);
            output.Write(enc.Push(plaintext.AsSpan(offset, len).ToArray()));
        }
        output.Write(enc.Finish());
        return output.ToArray();
    }

    /// <summary>
    /// Decrypt a media blob with the incremental stream API.
    ///
    /// <c>Finish</c> throws if the stream was truncated, so plaintext from
    /// <c>Push</c> must not be treated as complete until it succeeds.
    /// </summary>
    public byte[] DecryptMedia(byte[] ciphertext, byte[] conversationKey)
    {
        using var dec = _chat.StreamDecryptor(conversationKey);
        using var output = new MemoryStream();
        for (var offset = 0; offset < ciphertext.Length; offset += MediaChunk)
        {
            var len = Math.Min(MediaChunk, ciphertext.Length - offset);
            output.Write(dec.Push(ciphertext.AsSpan(offset, len).ToArray()));
        }
        output.Write(dec.Finish());
        return output.ToArray();
    }

    // -- Generic helpers (handy for metadata + tests) -----------------------

    public string Encrypt(string plaintext, byte[] conversationKey) => _chat.Encrypt(plaintext, conversationKey);

    public string Decrypt(string ciphertextB64, byte[] conversationKey) => _chat.Decrypt(ciphertextB64, conversationKey);

    public void Dispose() => _chat.Dispose();
}

/// <summary>Fields the X API expects for an encrypted message send.</summary>
public sealed record SendBody(string MessageId, string EncodedMessageCreateEvent, string EncodedMessageEventSignature);

/// <summary>Pull the plain text out of a decrypted Message event, or null.</summary>
public static class EventHelpers
{
    public static string? MessageText(JsonElement evt)
    {
        if (!evt.TryGetProperty("type", out var type) || type.GetString() != "Message")
            return null;
        if (evt.TryGetProperty("content", out var content) &&
            content.TryGetProperty("text", out var text))
            return text.GetString();
        return null;
    }

    /// <summary>
    /// Map a prepared conversation change into the X API request shape
    /// (snake_case field names, ready to serialize as the JSON body).
    ///
    /// Works for 1:1 key changes (one signature) and group create / member add
    /// (two signatures). <paramref name="signingPublicKey"/> is the sender's own
    /// signing key, which the API expects alongside each signature.
    /// </summary>
    public static Dictionary<string, object?> PrepToRequest(PreparedConversationChange prep, string signingPublicKey)
    {
        var actionSignatures = new List<Dictionary<string, object?>>();
        foreach (var sig in prep.ActionSignatures)
        {
            var entry = new Dictionary<string, object?>
            {
                ["message_id"] = sig.MessageId,
                ["encoded_message_event_detail"] = sig.EncodedMessageEventDetail,
                ["message_event_signature"] = new Dictionary<string, object?>
                {
                    ["signature"] = sig.Signature,
                    ["signature_version"] = sig.SignatureVersion,
                    ["public_key_version"] = sig.PublicKeyVersion,
                    ["signing_public_key"] = signingPublicKey,
                },
            };
            // Conversation-key-change payloads are withheld by the SDK (they
            // embed the plaintext key); only include the field when present.
            if (!string.IsNullOrEmpty(sig.SignaturePayload))
                entry["signature_payload"] = sig.SignaturePayload;
            actionSignatures.Add(entry);
        }
        return new Dictionary<string, object?>
        {
            ["conversation_key_version"] = prep.ConversationKeyVersion,
            ["conversation_participant_keys"] = prep.ParticipantKeys
                .Select(pk => new Dictionary<string, object?>
                {
                    ["user_id"] = pk.UserId,
                    ["encrypted_conversation_key"] = pk.EncryptedKey,
                    ["public_key_version"] = pk.PublicKeyVersion,
                })
                .ToList(),
            ["action_signatures"] = actionSignatures,
        };
    }
}
